### ISSUE TITLE
[DDC] if-case pattern does not match.

### ISSUE THREAD

**Issue by ykmnkmi** [dart-lang/sdk#54646]
This works with `dart2js`, but not with `ddc`:

```dart
import 'dart:math';

sealed class Node {}

final class Name extends Node {}

final class Text extends Node {}

void main() {
  var node = <Node>[Name(), Text()][Random().nextInt(2)];
  print(node);

  if (node case Name node) {
    print(node); // does not print when node is a `Name` instance
  }
}
```

Dart 3.2.3 (stable) on "windows_x64". Related ykmnkmi/jinja.dart#26.

**Comment by nshahan** [comment]
Looks like a duplicate of https://github.com/dart-lang/sdk/issues/54559. The fix has already landed and is currently in releases on the dev channel after [3.4.0-35.0.dev](https://github.com/dart-lang/sdk/releases/tag/3.4.0-35.0.dev)

A workaround would be to rename the local `node` variable that gets declared in the if case to avoid shadowing:
```dart
if (node case Name nameNode) {
    print(nameNode);
  }
```

@ykmnkmi can you comment on how blocking this issue is for you? Because the workaround is so easy there hasn't been any discussion to cherry pick this fix and right now it will naturally be released as part of Dart 3.4.0.

**Comment by nshahan** [comment]
Duplicate of https://github.com/dart-lang/sdk/issues/54559, forgot to close in the previous comment, but you can still respond here if needed.
